enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,747 @@
+
+```ini
+[crio.runtime.runtimes.high-performance]
+runtime_path = "/bin/crun"
+runtime_type = "oci"
+runtime_root = "/bin/crun"
+allowed_annotations = ["cpu-load-balancing.crio.io", "cpu-quota.crio.io", "irq-load-balancing.crio.io", "cpu-c-states.crio.io", "cpu-freq-governor.crio.io"]
+```
+
+#### Kubelet configuration 
+
+Because of multitude of option in Kubelet configuration, a simple passthrough (copy paste)
+will be implemented, rather than exposing every single little configuration variable.
+
+```yaml
+# /etc/microshift/config.yaml
+kubelet:
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"
+  cpuManagerReconcilePeriod: 5s
+  memoryManagerPolicy: Static
+  topologyManagerPolicy: single-numa-node
+  reservedSystemCPUs: 0,28-31
+  reservedMemory:
+  - limits:
+      memory: 1100Mi
+    numaNode: 0
+  kubeReserved:
+    memory: 500Mi
+  systemReserved:
+    memory: 500Mi
+  evictionHard:
+    imagefs.available: 15%
+    memory.available: 100Mi
+    nodefs.available: 10%
+    nodefs.inodesFree: 5%
+  evictionPressureTransitionPeriod: 0s
+```
+will be passed-through to kubelet config as:
+```yaml
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  full-pcpus-only: "true"
+cpuManagerReconcilePeriod: 5s
+memoryManagerPolicy: Static
+topologyManagerPolicy: single-numa-node
+reservedSystemCPUs: 0,28-31
+reservedMemory:
+- limits:
+    memory: 1100Mi
+  numaNode: 0
+kubeReserved:
+  memory: 500Mi
+systemReserved:
+  memory: 500Mi
+evictionHard:
+  imagefs.available: 15%
+  memory.available: 100Mi
+  nodefs.available: 10%
+  nodefs.inodesFree: 5%
+evictionPressureTransitionPeriod: 0s
+```
+
+#### Extra manifests
+
+Connects Pod's `.spec.runtimeClassName` to CRI-O's runtime.
+If Pod has `.spec.runtimeClassName: microshift-low-latency`,
+it can use annotations specified in CRI-O config with `crio.runtime.runtimes.high-performance`.
+
+```yaml
+apiVersion: node.k8s.io/v1
+handler: high-performance
+kind: RuntimeClass
+metadata:
+  name: microshift-low-latency
+```
+
+
+### Risks and Mitigations
+
+Biggest risk is system misconfiguration.
+It is not known to author of the enhancement if there are configurations (like kernel params, sysctl, etc.)
+that could brick the device, though it seems rather unlikely.
+Even if kernel panic occurs after staging a deployment with new configuration,
+thanks to greenboot functionality within the grub itself, the system will eventually rollback to
+previous deployment.
+Also, it is assumed that users are not pushing new image to production devices without prior verification on reference device.
+
+It may happen that some users need to use TuneD plugins that are not handled by the profile we'll create.
+In such case we may investigate if it's something generic enough to include, or we can instruct them
+to create new profile that would include `microshift-baseline` profile.
+
+Systemd daemon we'll provide to enable TuneD profile should have a strict requirement before it
```

Comment:
I'm not certain what you mean here - did you mean that "The Systemd daemon that enables the TuneD profile should have a strict requirement before it reboots the node to prevent it from entering a boot loop.” ?

Reply:
Yes, but the way of operation changed - it's up to user to say if we should reboot after applying the profile. I'll update it